File: 2021-kakao-blind-recruitment/Q2.메뉴-리뉴얼.py
from itertools import combinations
from collections import Counter


def solution(orders, course):
    menu_dic = {pick : [] for pick in course}
    ans = []
    
    for order in orders:
        for pick in course:
            for item in combinations(order, pick):
                voted_menu = ''.join(sorted(item))
                menu_dic[pick].append(voted_menu)

    for voted_menu_list in menu_dic.values():
        voted_menu_couter = Counter(voted_menu_list)
        
        if voted_menu_couter:
            most_voted_count = voted_menu_couter.most_common(1)[0][1]
            ans.extend([k for [k,v] in voted_menu_couter.items() if v == most_voted_count and most_voted_count > 1 ])
    
    return sorted(ans)

File: 2021-kakao-blind-recruitment/Q2.메뉴-리뉴얼.js
const cmp = (a, b) => a.localeCompare(b);

const sortString = text =>
	Array.from(text)
		.sort((a, b) => a.localeCompare(b))
		.join('');

class Counter {
	counter = new Map();

	constructor(arr) {
		arr.forEach(item => this.set(item));
	}

	get(key) {
		if (!this.counter.has(key)) new Error('key is not exist');
		return this.counter.get(key);
	}

	set(key) {
		if (!this.counter.has(key)) this.counter.set(key, 0);
		this.counter.set(key, this.counter.get(key) + 1);
	}

	mostCommon() {
		const maxCount = Math.max(...this.counter.values());
		return Array.from(this.counter.entries()).filter(
			([_, v]) => v === maxCount,
		);
	}
}

function combination(arr, pick) {
	let ans = [];

	const backtrack = (item, rest) => {
		if (item.length === pick) ans.push(item);
		Array.from(rest).forEach((v, i) => backtrack(item + v, rest.slice(i + 1)));
	};

	backtrack([], arr);
	return ans;
}

function solution(orders, course) {
	let ans = [];
	const menuMap = new Map(course.map(v => [v, new Array()]));
	const sortedOrders = orders.map(sortString);

	for (const order of sortedOrders) {
		for (const pick of course) {
			combination(order, pick).forEach(item => {
				menuMap.get(pick).push(item);
			});
		}
	}

	for (const items of menuMap.values()) {
		const counter = new Counter(items);
		const mostCommonItems = counter.mostCommon();

		if (!mostCommonItems.length || mostCommonItems[0][1] < 2) continue;
		mostCommonItems.forEach(item => ans.push(item[0]));
	}

	return ans.sort(cmp);
}
